File: solutions/python/code-carl/ch09_dp/t2_lc0070_climb_stairs.py
"""LeetCode 70. Climbing Stairs
https://leetcode.cn/problems/climbing-stairs/
"""


class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 1:
            return 1

        pre, curr = 1, 1
        for i in range(2, n + 1):
            pre, curr = curr, pre + curr

        return curr

if __name__ == "__main__":
    solution = Solution()
    assert solution.climbStairs(2) == 2
    assert solution.climbStairs(3) == 3
    print("All tests passed.")

File: solutions/code-carl/ch09_dp/T2_LC0070_ClimbStairs.java
/**
 * Practice template for:
 * - LeetCode 70. Climbing Stairs
 *   https://leetcode.cn/problems/climbing-stairs/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/20 15:10
 * @comment
 */
public class T2_LC0070_ClimbStairs {
    public int climbStairs(int n) {
        if (n <= 2) {
            return n;
        }
        int prepre = 1, pre = 2;
        for (int i = 3; i <= n; i++) {
            int curr = prepre + pre;
            prepre = pre;
            pre = curr;
        }
        return pre;
    }
}
